--- a/drivers/net/wireless/broadcom/b43/phy_ht.c
+++ b/drivers/net/wireless/broadcom/b43/phy_ht.c
@@ -567,6 +567,18 @@ static void b43_phy_ht_tx_power_ctl(stru
 					      phy_ht->tx_pwr_idx[i]);
 		for (i = 0; i < 3; i++)
 			b43_phy_set(dev, cmd_regs[i], en_bits);
+
+		if (b43_cg3000_4331_board(dev)) {
+			for (i = 0; i < 3; i++) {
+				u16 init = b43_phy_read(dev, cmd_regs[i]) &
+					   B43_PHY_HT_TXPCTL_CMD_C1_INIT;
+
+				b43_phy_maskset(dev, cmd_regs[i],
+						~(B43_PHY_HT_TXPCTL_CMD_C1_INIT |
+						  en_bits) & 0xffff,
+						init | en_bits);
+			}
+		}
 	}
 
 	phy_ht->tx_pwr_ctl = enable;
