--- a/drivers/net/wireless/broadcom/b43/main.c
+++ b/drivers/net/wireless/broadcom/b43/main.c
@@ -1918,12 +1918,16 @@ static void b43_write_beacon_template(st
 	/* We can't send beacons with short preamble. Would get PHY errors. */
 	ctl &= ~B43_TXH_PHY_SHORTPRMBL;
 	ctl &= ~B43_TXH_PHY_ANT;
+	if (b43_cg3000_4331_board(dev))
+		ctl &= ~B43_TXH_PHY_TXPWR;
 	ctl &= ~B43_TXH_PHY_ENC;
 	ctl |= antenna;
 	if (b43_is_cck_rate(rate))
 		ctl |= B43_TXH_PHY_ENC_CCK;
 	else
 		ctl |= B43_TXH_PHY_ENC_OFDM;
+	if (b43_cg3000_4331_board(dev))
+		ctl |= b43_cg3000_txpwr_phy_ctl(dev);
 	b43_shm_write16(dev, B43_SHM_SHARED, B43_SHM_SH_BEACPHYCTL, ctl);
 
 	if (dev->phy.type == B43_PHYTYPE_HT) {
@@ -3593,11 +3603,19 @@ static void b43_mgmtframe_txantenna(stru
 
 	/* For ACK/CTS */
 	tmp = b43_shm_read16(dev, B43_SHM_SHARED, B43_SHM_SH_ACKCTSPHYCTL);
+	if (b43_cg3000_4331_board(dev))
+		tmp &= ~B43_TXH_PHY_TXPWR;
 	tmp = (tmp & ~B43_TXH_PHY_ANT) | ant;
+	if (b43_cg3000_4331_board(dev))
+		tmp |= b43_cg3000_txpwr_phy_ctl(dev);
 	b43_shm_write16(dev, B43_SHM_SHARED, B43_SHM_SH_ACKCTSPHYCTL, tmp);
 	/* For Probe Resposes */
 	tmp = b43_shm_read16(dev, B43_SHM_SHARED, B43_SHM_SH_PRPHYCTL);
+	if (b43_cg3000_4331_board(dev))
+		tmp &= ~B43_TXH_PHY_TXPWR;
 	tmp = (tmp & ~B43_TXH_PHY_ANT) | ant;
+	if (b43_cg3000_4331_board(dev))
+		tmp |= b43_cg3000_txpwr_phy_ctl(dev);
 	b43_shm_write16(dev, B43_SHM_SHARED, B43_SHM_SH_PRPHYCTL, tmp);
 }
 
