--- a/drivers/net/wireless/broadcom/b43/phy_ht.c
+++ b/drivers/net/wireless/broadcom/b43/phy_ht.c
@@ -1511,24 +1511,24 @@ static void b43_phy_ht_cg3000_apply_aa2
 
 	switch (aa2g) {
 	case 1:
-		b43_phy_maskset(dev, 0x0410, ~0x0002, 0x0000);
-		b43_phy_maskset(dev, 0x0410, ~0x0001, 0x0000);
-		b43_phy_maskset(dev, 0x04b0, ~0x0020, 0x0020);
-		b43_phy_maskset(dev, 0x04b1, ~0x0200, 0x0000);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x0410, ~0x0002, 0x0000);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x0410, ~0x0001, 0x0000);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x04b0, ~0x0020, 0x0020);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x04b1, ~0x0200, 0x0000);
 		break;
 	case 2:
-		b43_phy_maskset(dev, 0x0410, ~0x0002, 0x0000);
-		b43_phy_maskset(dev, 0x0410, ~0x0001, 0x0001);
-		b43_phy_maskset(dev, 0x04b0, ~0x0020, 0x0020);
-		b43_phy_maskset(dev, 0x04b1, ~0x0200, 0x0200);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x0410, ~0x0002, 0x0000);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x0410, ~0x0001, 0x0001);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x04b0, ~0x0020, 0x0020);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x04b1, ~0x0200, 0x0200);
 		break;
 	case 3:
-		b43_phy_maskset(dev, 0x0410, ~0x0002, 0x0002);
-		b43_phy_maskset(dev, 0x0410, ~0x0001, 0x0000);
-		b43_phy_maskset(dev, 0x04b0, ~0x0020, 0x0020);
-		b43_phy_maskset(dev, 0x04b1, ~0x0200, 0x0000);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x0410, ~0x0002, 0x0002);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x0410, ~0x0001, 0x0000);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x04b0, ~0x0020, 0x0020);
+		b43_phy_ht_cg3000_phy_maskset_direct(dev, 0x04b1, ~0x0200, 0x0000);
 		hostf5 |= B43_HF5_CG3000_AA2G3;
 		break;
 	default:
